### DISCUSSION TITLE
Idea: Javascript Viewer

### DISCUSSION THREAD

**Discussion by GoldcatGames** [ScratchAddons/ScratchAddons#3870] [Ideas]
I Got This Idea From The Coding Platform I Started On Before Scratch (Tynker BTW) And I Remember That There Was A JavaScript Button Where It Would Basically Show What The Block Coding Would Look Like As JavaScript, And That Would Be Really Helpful For People Who Are Trying To Learn Javascript Too.

**Comment by Samq64** [comment]
Related discussion: #3728

**Reply by GoldcatGames** [reply]
that's not what I meant, I didn't mean a converter, I meant not editable javascript, you basically just click a button, then it shows javascript code, that you can just see and that's it.

**Comment by Dhuls-Scratch** [comment]
https://leopardjs.com/

**Reply by GoldcatGames** [reply]
that's not what I meant, I didn't mean a converter, I meant not editable javascript, you basically just click a button, then it shows javascript code, that you can just see and that's it.

**Comment by Dhuls-Scratch** [comment]
The JavaScript Scratch uses is a bit weird and wouldn’t be helpful to anyone. In fact, a simple script to forever move a “my variable” amount of steps would look like this:
````
{
  "va[U{Cbi_NZpSOSx_kVA": {
    "opcode": "event_whenflagclicked",
    "inputs": {},
    "fields": {},
    "next": "tzXnZ{8G!xK|t^WAWF{m",
    "topLevel": true
  },
  "tzXnZ{8G!xK|t^WAWF{m": {
    "opcode": "control_forever",
    "inputs": {
      "SUBSTACK": {
        "name": "SUBSTACK",
        "block": "$xf$bq|xl(}RhT-K,taS"
      }
    },
    "fields": {},
    "next": null,
    "topLevel": false
  },
  "$xf$bq|xl(}RhT-K,taS": {
    "opcode": "motion_movesteps",
    "inputs": {
      "STEPS": {
        "name": "STEPS",
        "block": "cw__.I:g}Y~`:5KmO00q"
      }
    },
    "fields": {},
    "next": null,
    "topLevel": false
  },
  "cw__.I:g}Y~`:5KmO00q": {
    "opcode": "data_variable",
    "inputs": {},
    "fields": {
      "VARIABLE": {
        "name": "VARIABLE",
        "id": ***@***.***|i[#Fk?(8x)AV.-my variable"
      }
    },
    "next": null,
    "topLevel": false
  }
}
````

**Reply by CST1229** [reply]
That's the abstract syntax tree. Scratch doesn't really convert anything into JavaScript.

**Reply by Secret-chest** [reply]
What about scratch-vm?

**Reply by GrahamSH-LLK** [reply]
> What about scratch-vm?

State != JS
There is an internal state represented in objects/arrays etc, but your code is never really "in" javascript. It is being executed _by_ javascript.

**Comment by Secret-chest** [comment]
You can't do it without:
a) writing code for all blocks first (like I do in [Scratch2Python](https://github.com/Secret-chest/scratch2python), even though it is python, you get it)
b) importing a Scratch library (scratch-vm) which is very hard in js, you have to deal with npm and all this stuff...

**Reply by GrahamSH-LLK** [reply]
Scratch-vm is not what you think it is 😄 

It's an internal package scratch uses to represent, run, and maintain the state of scratch programs. SA automatically exposes the VM to addons, and you could (in theory) convert blocks to js using the VM and blockly.
